### **Soal 1 Pandas Data Frame**

In [95]:
import pandas as pd

In [96]:
from sklearn.datasets import make_regression
# generate dataset
X, y = make_regression(n_samples=100, n_features=100, n_informative=10)

In [97]:
data_X = pd.DataFrame(X,columns = ['X'+str(i) for i in range(X.shape[1])])

In [98]:
data_y = pd.DataFrame(y,columns = ['Y'])

In [99]:
data_X['Y'] = data_y

In [100]:
data_X

X0        X1        X2        X3        X4        X5        X6  \
0   0.955668  1.078449  0.343145 -1.892888  0.040148 -0.350689 -1.625874   
1  -0.173052  0.287296 -1.640557 -0.489669 -1.862366 -2.274201  0.736967   
2   0.420153  0.284085  1.580256 -1.115777 -1.435245  0.493722  1.069172   
3  -0.358731  1.004772 -2.266371 -1.170504 -0.402933  0.929559  0.505864   
4   0.951641 -0.696266  0.500442  2.090788  1.782319 -1.840191  0.535899   
..       ...       ...       ...       ...       ...       ...       ...   
95 -2.315042 -0.072743  1.733087 -1.266464  1.832773  0.720813 -0.159121   
96 -0.112857  0.319268 -0.555037 -1.275811  0.273468  1.065628 -0.785600   
97  1.674983  0.222766  0.518243 -1.511081 -0.265491  0.343453  1.559395   
98 -0.267647  0.876875  0.658478 -1.252501  1.862551 -0.682050  1.046784   
99  0.423173 -1.163834 -2.055630 -1.649208 -0.999351  0.126458  1.180408   

          X7        X8        X9  ...       X91       X92       X93       X94  \
0  -0.391149  0.826271 -1.324590  ... -0.240450  0.228462  0.780853 -1.283797   
1   0.366445 -0.954375 -0.301500  ... -0.055826  0.788664  0.630684 -0.768037   
2   0.114474 -0.074178 -0.805477  ... -0.387580 -0.908251  1.302724  0.305619   
3  -0.358738 -0.850176 -1.323721  ... -1.322462 -1.963581 -1.760075  0.389868   
4   0.428536 -1.003720 -0.945001  ... -0.543537 -1.292224 -0.086940  0.562970   
..       ...       ...       ...  ...       ...       ...       ...       ...   
95  0.462551  1.061113  0.043362  ... -0.573372 -0.832661 -0.459934 -0.977755   
96  1.451373  2.684106 -0.761739  ...  0.478370 -0.207178 -1.723993  0.090671   
97  0.044012 -1.166825 -0.009573  ...  0.925814  2.269468 -0.626861  0.219867   
98 -1.208177  0.435716  0.863181  ... -0.693294  0.723318 -2.087796 -2.641431   
99  0.271413  2.337027  0.473022  ... -0.479485  0.446794 -0.581543  0.276559   

         X95       X96       X97       X98       X99           Y  
0  -0.152387 -0.659632 -0.139578 -0.816379  1.401141  104.884683  
1   0.787812  0.613591 -1.017883  1.041246  2.372351  -46.567454  
2  -0.380964  1.278645 -1.464784 -0.520885  0.077197  162.266838  
3  -1.262042  0.402504  1.013935 -0.924977  0.373802 -151.255168  
4   1.747710 -1.760810  1.348671  1.689289  1.369069   37.738189  
..       ...       ...       ...       ...       ...         ...  
95 -2.205244  1.540907 -1.191674 -0.649013  0.535099 -119.420110  
96  0.360089  1.451574  0.669370  0.762799 -0.242177   69.104283  
97  0.469130  0.453630 -0.217314  0.493793  0.067476 -120.818446  
98 -1.102907 -0.866145  0.066233  0.651673  1.065423   36.150068  
99 -0.353688 -1.942759 -1.199667 -2.492590  0.703014   -5.135058  

[100 rows x 101 columns]

Buatlah sebuah Dataframe dengan data di atas. Dimana terdapat variabel X dan y. buatlah menjadi column X0-X99 dan y pada akhir column.

In [101]:
#jangan di run
#cell contoh hasil

### Soal 02

Berapa banyak perenang yang memiliki medali perunggu(bronze)?

In [102]:
import pandas as pd
df=pd.read_csv('summer.csv')
df

Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender                       Event  \
0                  HAJOS, Alfred     HUN    Men              100M Freestyle   
1               HERSCHMANN, Otto     AUT    Men              100M Freestyle   
2              DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors   
3             MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors   
4             CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors   
...                          ...     ...    ...                         ...   
31160         JANIKOWSKI, Damian     POL    Men                    Wg 84 KG   
31161  REZAEI, Ghasem Gholamreza     IRI    Men                    Wg 96 KG   
31162             TOTROV, Rustam     RUS    Men                    Wg 96 KG   
31163          ALEKSANYAN, Artur     ARM    Men                    Wg 96 KG   
31164             LIDBERG, Jimmy     SWE    Men                    Wg 96 KG   

        Medal  
0        Gold  
1      Silver  
2      Bronze  
3        Gold  
4      Silver  
...       ...  
31160  Bronze  
31161    Gold  
31162  Silver  
31163  Bronze  
31164  Bronze  

[31165 rows x 9 columns]

In [103]:
#your code goes here
df_coba = df.filter(items=['Discipline','Medal'])

In [104]:
df_coba

Discipline   Medal
0                 Swimming    Gold
1                 Swimming  Silver
2                 Swimming  Bronze
3                 Swimming    Gold
4                 Swimming  Silver
...                    ...     ...
31160  Wrestling Freestyle  Bronze
31161  Wrestling Freestyle    Gold
31162  Wrestling Freestyle  Silver
31163  Wrestling Freestyle  Bronze
31164  Wrestling Freestyle  Bronze

[31165 rows x 2 columns]

In [114]:
tes = df_coba.value_counts()

In [117]:
pd.options.display.max_rows=250

In [118]:
display(tes)

Discipline             Medal 
Athletics              Silver    1224
                       Gold      1215
                       Bronze    1199
Swimming               Gold       901
Rowing                 Gold       890
                       Silver     889
                       Bronze     888
Swimming               Silver     876
                       Bronze     851
Artistic G.            Gold       752
                       Silver     688
                       Bronze     663
Fencing                Gold       552
                       Silver     542
                       Bronze     519
Football               Bronze     509
                       Gold       503
Hockey                 Gold       486
Football               Silver     485
Hockey                 Silver     477
                       Bronze     459
Boxing                 Bronze     416
Sailing                Gold       412
Shooting               Silver     386
                       Gold       384
                    

Answer:
**perenang yang memiliki medali perunggu(bronze) ada 851**

Dalam olahraga air (Aquatics), kota mana yang memiliki rata-rata perolehan medali tertinggi?

In [4]:
#your code goes here

Answer:

### **Soal 3 List Comprehension & Pandas** 

Pada Soal ini kalian **wajib** menggunakan feature list comprehension dan juga apply.<br>

In [74]:
import pandas as pd
df=pd.read_csv('Salaries.csv')
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                    Not provided        NaN   
148651                                    Not provided        NaN   
148652                                    Not provided        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

PT.A baru saja kena tegur dari negara karena ketauan tidak membayar uang lemburan. Anda sebagai seorang data Scientist diminta PT.A untuk mengolah data yang ada.

**Step pertama**

**Wajib Menggunakan .Apply()** <br>
PT.A melihat beberapa data tidak di isi oleh sebagian karyawan. kalian di minta untuk untuk merubah nilai 'Not provided' pada column `JobTitle` dan `EmployeeName` menjadi 'Error'.

In [75]:
#your code goes here
df['JobTitle'] = df['JobTitle'].replace(['Not provided'], 'Error')

In [76]:
df['EmployeeName'] = df['EmployeeName'].replace(['Not provided'], 'Error')

In [28]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

**Step Kedua**

**WAJIB Gunakan list comprehension**<br>
Anda diminta untuk menambahkan 200000 pada karyawan yang mendapat Overtime 0 pada Data yang telah dibersihkan di tahap 1.


In [ ]:
#your code goes here

In [85]:
df['OvertimePay'] = df['OvertimePay'].replace(0, 200000)

In [86]:
df.tail()

Id   EmployeeName                    JobTitle  BasePay  \
148649  148650  Roy I Tillery                   Custodian      0.0   
148650  148651          Error                       Error      NaN   
148651  148652          Error                       Error      NaN   
148652  148653          Error                       Error      NaN   
148653  148654      Joe Lopez  Counselor, Log Cabin Ranch      0.0   

        OvertimePay  OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  \
148649     200000.0      0.00       0.0      0.00              0.00  2014   
148650          NaN       NaN       NaN      0.00              0.00  2014   
148651          NaN       NaN       NaN      0.00              0.00  2014   
148652          NaN       NaN       NaN      0.00              0.00  2014   
148653     200000.0   -618.13       0.0   -618.13           -618.13  2014   

        Notes         Agency  Status  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN

In [29]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         200000.00  400184.25       NaN  567595.43         567595.43  2011   
1         122565.94  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649    200000.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653    200000.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

**Tahap ketiga** 

**bebas memakai teknik apapun**<br>
selanjutnya ada kelasahan pada penulisan 'CAPTAIN III (POLICE DEPARTMENT)' ubahlah menjadi 'CAPTAIN III' saja

In [ ]:
#your code goes here

In [87]:
df['JobTitle'] = df['JobTitle'].replace(['CAPTAIN III (POLICE DEPARTMENT)'], 'CAPTAIN III')

In [30]:
#jangan tulis code disini
#cell contoh hasil:

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651              Error   
148651  148652              Error   
148652  148653              Error   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                                          CAPTAIN III  155966.02   
2                                          CAPTAIN III  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                           Error        NaN   
148651                                           Error        NaN   
148652                                           Error        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         200000.00  400184.25       NaN  567595.43         567595.43  2011   
1         122565.94  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649    200000.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653    200000.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

### **Soal 4 One Hot Encoding**

Buatlah **sebuah fungsi** bernama `one_hot`, yang menerima input sebuah numpy array.

Ex:<br>
input array [2,3,1,2] maka hasil dari fungsi kalian:

array [[0,1,0], [0,0,1], [1,0,0], [0,1,0]]

Constrain: <br>

- Integers dalam input selalu di awali 1. 
- Integers dalam input hanya positif dan bukan 0.
- Integers dalam input tidak akan skip angka.

In [37]:
import numpy as np
input_array = np.array([3,2,1,2,3,3,3])
#your code goes here
def one_hot ():

In [3]:
#jangan tulis code disini
#cell contoh hasil:

array([[0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.]])